examples/khr-3hv/khr-3hv_continuous/controllers/supervisorsManager/khr3hvEnv.py

```diff
@@ -0,0 +1,206 @@
+                       "Current x position": self.robot.getPosition()[0],
+                       "Globan distance walekd": GLOBAL_DISTANCE
+                       })
+
+        motorPos.extend([i for i in self.motorPositionArr])
+        # Shift the last 10 observations by one, on the time window of 5
+        self.obs = shift(self.obs, -self.numObs, cval=np.NaN)
+        self.obs[-self.numObs:] = motorPos
+
+        return np.array(self.obs)
+
+    def get_reward(self, action):
+        """
+        Calculate the reward of the agent. The reward award agent when moving forward on the x-axis.
+        Also, we are using a second term so the agent moves further away from his previous position.
+        :return: Reward value
+        :rtype: float
+
+        """
+        reward = 2.5 * \
+            self.robot.getPosition()[0] + \
+            self.robot.getPosition()[0] - self.prev_pos
+
+        if self.counterLoging % self.save_freq == 0 and self.train and not self.useRay:
+            wandb.log({"reward": reward,
+                       "reward-1term-weight-pos": 2.5 * self.robot.getPosition()[0],
+                       "reward-2term-diff-possition": self.robot.getPosition()[0] - self.prev_pos
+                       })
+
+        self.prev_pos = self.robot.getPosition()[0]
+        return reward
+
+    def is_done(self):
+        """
+        This method checks the termination criteria for each episode. 
+        If the criteria are satisfied return True otherwise it returns False.
+        :return: Termination criteria
+        :rtype: Boolean
+
+        """
+        # Agent fall.
+        if self.robot.getPosition()[2] < 0.5:
+            return True
+        # Agent walked out of the box
+        if self.robot.getPosition()[0] > 15:
+            return True
+        return False
+
+    def get_default_observation(self):
+        """
+        Simple implementation returning the default observation which is a zero 
+        vector in the shape of the observation space.
+        :return: Starting observation zero vector
+        :rtype: numpy array
+        """
+        self.prev_pos = 0
+        self.obs = np.zeros(self.t * self.numObs)
+        self.distance_walk = 0
+
+        return np.zeros(self.observation_space.shape[0])
+
+    def apply_action(self, action):
+        """
+        This method uses the action list provided, which contains the next action to be 
+        executed by the robot.It contains a float number denoting the action. 
+        The corresponding motorPosition value is applied at each motor.
+        :param action: The list that contains the action value
+        :type action: list of float
+        """
+        motorIndexes = [0, 1, 2, 3, 4, 5]
+
+        for i, ac in zip(motorIndexes, action):
+            self.motorPositionArr[i] += ac
+            self.motorList[i].setVelocity(self.motor_velocity)
+            self.motorList[i].setPosition(ac)
```

Comment:
I am not exactly sure i can see the "not synchronized" comment you made for this line (see edit at the bottom). As far as i understand, the intention behind this is to save the motor positions in `motorPositionArr` which starts at zeros, so it can be used in the observation, so they are adding the action `ac` to `motorPositionArr`. Apparently this also happens in `.setPosition(ac)`, if the starting motor position is at zero.

What concerns me most is the way the motors are generally handled here, setting both the a desired velocity as well as a position. According to [this](https://cyberbotics.com/doc/reference/motor#motor-control-summary) and [this](https://cyberbotics.com/doc/reference/motor#motor-control-summary), you only need to set the position and let the PID controller handle it. 

Edit: apparently the position is initially set to `inf` in the `getAllMotors` util function so i don't really know how it behaves here, and whether indeed the actual position of the motor is the same as in the `motorPositionArr`. You are probably right about the not synchronized comment. It would be better to set `motorPositionArr` from the [getTargetPosition](https://cyberbotics.com/doc/reference/motor?tab-language=python#wb_motor_get_target_position) method (see description lower).

Reply:
The thing you mentioned is similar to what I set for panda [here](https://github.com/aidudezzz/deepworlds/blob/5e815cf829eea46e05e82d8819cc3d755b9f909a/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_ddpg.py#L88). I use [`getValue()`](https://github.com/aidudezzz/deepworlds/blob/5e815cf829eea46e05e82d8819cc3d755b9f909a/examples/panda/panda_goal_reaching/controllers/ArmUtil.py#L18) to check if the motor reaches its target position in the simulator. That is also a problem!

I want to point out that `motorPositionArr` should be the same as [`what we pass to setPosition()`](https://github.com/KelvinYang0320/deepworlds/blob/db9ca33323020778e175b6aa141a7cd9b7179463/examples/khr-3hv/khr-3hv_continuous/controllers/supervisorsManager/khr3hvEnv.py#L183) because [this line](https://github.com/KelvinYang0320/deepworlds/blob/db9ca33323020778e175b6aa141a7cd9b7179463/examples/khr-3hv/khr-3hv_continuous/controllers/supervisorsManager/khr3hvEnv.py#L113) in `get_observations()`.

There are two possible solutions:
1. `self.motorPositionArr[i] += ac` → `self.motorPositionArr[i] = ac`
2. `self.motorList[i].setPosition(ac)`→`self.motorList[i].setPosition(self.motorPositionArr[i])` ([current ver. in our khr-3hv example](https://github.com/aidudezzz/deepworlds/blob/5e815cf829eea46e05e82d8819cc3d755b9f909a/examples/khr-3hv/controllers/supervisors_controller/supervisors_controller.py#L143-L158))

Reply:
Yeah i remember us having a pretty similar conversation some time ago for Panda actions and movement :smile:. You can go ahead and fix that the way you think that is best, either here or in a new PR later. Feel free to resolve this conversation whenever you want.

Reply:
@tsampazk I think we can fix that in another new PR later, so I have opened a new issue to fix these training problems.